Math/NumberTheory/ArithmeticFunctions/NFreedom.hs

```diff
@@ -0,0 +1,51 @@
+-- |
+-- Module:      Math.NumberTheory.ArithmeticFunctions.NFreedom
+-- Copyright:   (c) 2018 Alexandre Rodrigues Baldé
+-- Licence:     MIT
+-- Maintainer:  Alexandre Rodrigues Baldé <alexandrer_b@outlook.com>
+--
+-- N-free number generation.
+--
+
+module Math.NumberTheory.ArithmeticFunctions.NFreedom
+  ( sieveBlockNFree
+  ) where
+
+import Control.Monad                        (forM_)
+import Control.Monad.ST                     (runST)
+import qualified Data.Vector.Unboxed         as U
+import qualified Data.Vector.Unboxed.Mutable as MU
+
+import Math.NumberTheory.Primes             (primes, unPrime)
+import Math.NumberTheory.Utils.FromIntegral (wordToInt)
+
+-- | Evaluate the `isNFreeA` function over a block.
+-- Value of @f@ at 0, if zero falls into block, is undefined.
+--
+-- >>> sieveBlockNFree 2 1 10
+-- [True, True, True, False, True, True, True, False, False, True]
+sieveBlockNFree
+  :: Word
+  -> Word
+  -> Word
+  -> U.Vector Bool
+sieveBlockNFree _ _ 0 = U.empty
+sieveBlockNFree n lowIndex' len'
+  = runST $ do
+    as <- MU.replicate len True
+    forM_ ps $ \p -> do
+      let pPow = p ^ n
+          offset = negate lowIndex `mod` pPow
+      forM_ [offset, offset + pPow .. len - 1] $ \ix -> do
+          MU.unsafeWrite as ix False
+    U.unsafeFreeze as
+
+  where
+    lowIndex :: Int
+    lowIndex = wordToInt lowIndex'
+
+    len :: Int
+    len = wordToInt len'
+
+    ps :: [Int]
+    ps = take len $ map unPrime primes
```

Comment:
There's also this incorrect line. I did this because otherwise the function wouldn't terminate, but this'll always choose the same primes and cause `n`-unfree numbers to be incorrectly reported as free if one of their factors is after the 256th prime in `primes.`

This should probably involve `lowIndex` as well.

Reply:
Fixed.